### PR TITLE
Feat/unified gateway routing with per-agent control

### PR DESCRIPTION
**Pull request by kinglegendzzh**
# feat: Unified AI Gateway Routing with per-agent control

## Thinking Path

> - Paperclip orchestrates AI agents for zero-human companies
> - Each agent is bound to a single adapter (Gemini, Codex, OpenCode, etc.), and each adapter talks to a specific AI vendor
> - But AI vendors impose their own rate limits, per-token billing, and availability SLAs — when a vendor is down or quota is exhausted, the agent simply fails
> - Teams running fleets of agents across multiple pay-as-you-go providers (Alibaba Bailian, Volcengine, SiliconFlow, OpenRouter, etc.) need precise cost control and automatic failover without manual intervention
> -  ( https://github.com/paperclipai/paperclip/pull/2738 ) The `fallbackModels` field only works within a single adapter— it cannot route to a completely different adapter type (e.g., fall back from Gemini to Codex)
> - This pull request adds a **Unified AI Gateway** layer that sits between the Heartbeat scheduler and adapter execution, providing priority-based cross-adapter routing, per-route token/request quotas, and circuit breaker protection
> - It also includes a per-agent toggle to opt in/out of gateway routing, and a full management UI integrated into both the sidebar and agent detail views
> - The benefit is that operators can define a single routing policy (e.g., Gemini → Codex → OpenCode) that applies transparently to any agent, with automatic failover, quota enforcement, and cost control — while the existing budget management system continues to track all token consumption regardless of which route is active

### New Files (13)

| File | Purpose |
|---|---|
| `packages/db/src/schema/gateway_routes.ts` | Drizzle schema for `gateway_routes` table — route config with adapter type, model, priority, weight, quota limits, circuit breaker settings |
| `packages/db/src/schema/gateway_usage_counters.ts` | Drizzle schema for `gateway_usage_counters` — sliding-window token/request counters per minute/hour/day with unique index for upserts |
| `packages/db/src/schema/gateway_circuit_state.ts` | Drizzle schema for `gateway_circuit_state` — persistent circuit breaker state surviving server restarts |
| `packages/shared/src/types/gateway.ts` | Shared TypeScript types: `GatewayRoute`, `GatewayRouteHealth`, `GatewayRouteWithHealth`, `GatewayCircuitBreakerState` |
| `server/src/services/gateway.ts` | Core gateway service: route resolution (priority → availability → weighted random with agent-specific preference), in-memory rate limiting with DB-backed upsert persistence, atomic circuit breaker state machine, CRUD operations, stale counter purge |
| `server/src/routes/gateway.ts` | REST API endpoints with `assertAgentOwnership` security guard: `GET/POST/PATCH/DELETE /companies/:id/gateway/routes`, `GET /health`, `POST /reset-circuit` |
| `ui/src/api/gateway.ts` | Frontend API client for all gateway CRUD + health + circuit reset operations |
| `ui/src/pages/GatewayRouting.tsx` | Full gateway management page: stats tiles, route cards with **inline editing**, real-time usage bars, circuit breaker badges. Create/edit form shares `RouteFormState` interface + `RouteFormFields` component — zero code duplication between create and edit flows |
| `ui/src/components/TagInput.tsx` | Reusable tag input component with Enter-to-add, drag-to-reorder, and dropdown suggestions — used for per-route `fallbackModels` configuration |
| `ui/src/components/EnvVarEditor.tsx` | **Extracted** from `AgentConfigForm` into its own module — reusable environment variable editor supporting plain values and Secret refs (`secret_ref` with Seal-to-vault) |
| `doc/GATEWAY.md` | Feature documentation: concepts, API reference, configuration guide |
| `server/src/__tests__/gateway-service.test.ts` | Unit tests: 20 tests covering route resolution (including same-priority agent preference and company-wide fallback), rate limiting, circuit breaker, health reporting, edge cases |
| `packages/db/src/migrations/0054_adapter_config_overrides.sql` | DDL adding `adapter_config_overrides` JSONB column to `gateway_routes` |

### Modified Files (19)

| File | Change |
|---|---|
| `packages/db/src/schema/index.ts` | Export 3 new gateway tables |
| `packages/db/src/migrations/meta/_journal.json` | Register migrations `0053_graceful_speed.sql` and `0054_adapter_config_overrides.sql` (renumbered from 0050/0051 to avoid collisions after upstream merge) |
| `packages/db/src/migrations/0053_graceful_speed.sql` | DDL for `gateway_routes`, `gateway_usage_counters`, `gateway_circuit_state` |
| `packages/shared/src/types/index.ts` | Export gateway types |
| `packages/shared/src/index.ts` | Re-export gateway types from package root |
| `server/src/app.ts` | Mount gateway router |
| `server/src/routes/index.ts` | Register gateway routes |
| `server/src/index.ts` | **Startup**: load persisted circuit breaker states after DB init. **Periodic**: hourly-throttled stale usage counter purge (7-day retention) |
| `server/src/services/heartbeat.ts` | **Core integration**: resolve gateway route before `adapter.execute()` with `.catch()` graceful degradation, apply adapter/model/timeout overrides, record success/failure for counters and circuit breaker. **Bug fix**: clear incompatible `fallbackModels` during cross-adapter routing. **Per-agent bypass**: check `agent.metadata.gatewayBypass` flag. **Safety**: use `effectiveAdapterType` for `persistAdapterManagedRuntimeServices` |
| `server/src/services/agents.ts` | **FK safety**: delete `gatewayRoutes` referencing agent before agent deletion in `remove()` transaction |
| `server/src/services/companies.ts` | **FK safety**: delete `gatewayRoutes` by `companyId` before agents/company deletion in `remove()` transaction |
| `ui/src/App.tsx` | Register `GatewayRouting` page route |
| `ui/src/components/Sidebar.tsx` | Add "Gateway" nav item |
| `ui/src/components/AgentConfigForm.tsx` | **Major UX overhaul**: "Actual Execution Chain" banner showing priority routing, per-agent On/Off toggle, section headers relabeled to "Last Resort" when gateway is active. When gateway is active, fallback models field is replaced by a blue info banner linking to per-route gateway config. **Exported** `AdapterTypeDropdown` for reuse by Gateway UI |
| `ui/src/pages/AgentDetail.tsx` | Add "Gateway" tab with per-agent route management |
| `ui/src/lib/queryKeys.ts` | Add gateway query keys |
| `ui/src/lib/company-routes.ts` | Add `"gateway"` to `BOARD_ROUTE_ROOTS` |
| `packages/adapters/gemini-local/src/server/parse.ts` | Minor: related adapter fix |
| `package.json` | Dependency housekeeping |

### Key Design Decisions

1. **Transparent by default** — If no gateway routes exist, agents behave exactly as before. Zero config change required for existing setups.

2. **Cross-adapter routing** — The gateway can route an `opencode_local` agent to `gemini_local` or `codex_local` at runtime. When this happens, the agent's legacy `fallbackModels` are automatically cleared because they belong to the original adapter's model namespace (prevents `ModelNotFoundError`). When gateway is active, the AgentConfigForm replaces the fallback models field with a gateway-managed info banner, directing users to configure per-route fallback chains instead.

3. **Per-agent opt-out** — Each agent can individually bypass gateway routing via `metadata.gatewayBypass = true`, stored in the existing JSONB `metadata` column (no schema migration needed). The UI provides an instant On/Off toggle.

4. **Graceful degradation** — If the `gateway_routes` table doesn't exist (migration not yet applied) or a transient DB error occurs, `resolveRoute` is wrapped in `.catch()` and falls back to the agent's default adapter. A misconfigured deployment will never crash agent execution.

5. **Atomic persistence** — Both `persistCircuitState` and `persistUsageCounter` use `INSERT ... ON CONFLICT DO UPDATE` upserts to avoid race conditions between concurrent heartbeats. The earlier read-then-write pattern was replaced after code review.

6. **Company-scoped security** — `assertAgentOwnership()` validates that any `agentId` supplied to route creation/update belongs to the target `companyId`, preventing cross-company agent scoping via crafted API requests.

7. **FK cascade safety** — Agent and company deletion transactions now explicitly delete `gatewayRoutes` rows before the parent entity, preventing FK constraint violations from the `ON DELETE no action` policy.

8. **Budget integration** — Gateway routing is budget-aware: the native budget system tracks token consumption regardless of which route is active. Future work will tighten this integration for hard-stop enforcement at the gateway level.

9. **In-memory rate limiting with DB observability** — Sliding window counters are held in memory for low-latency route decisions. Counter snapshots are upserted to DB for observability and survive restarts as historical data. Stale rows (>7 days) are purged hourly. This is suitable for single-instance deployments; Redis would be needed for multi-instance production.

10. **Route-Level Adapter Configuration** — Gateway routes can specify `adapterConfigOverrides` (stored as JSONB) to customize environment variables, extra arguments, max turns, thinking effort, and **per-route fallback models** for the underlying adapter without modifying the global agent settings.

11. **Non-invasive UI component reuse** — The gateway route form imports `AdapterTypeDropdown`, `EnvVarEditor`, `Field`, `DraftInput`, `DraftNumberInput`, and `TagInput` from shared modules. `EnvVarEditor` has been extracted into its own file (`ui/src/components/EnvVarEditor.tsx`) and `TagInput` is a new reusable component. When the adapter registry adds new types or these shared components gain new functionality, the gateway form automatically inherits the changes — zero code duplication.

12. **Merged candidate pool with agent-specific preference** — Agent-specific and company-wide routes are merged into a single candidate pool, grouped by priority. **Within each priority tier, agent-specific routes are preferred** over company-wide ones. When all agent-specific routes at a given priority are exhausted or circuit-tripped, company-wide routes at the same or lower priority are consulted. Operators can set company-wide safety-net routes and trust they will be used as fallbacks.

13. **Shared form architecture (Create/Edit)** — Route create and inline-edit forms share a single `RouteFormState` interface and `RouteFormFields` component, eliminating duplication and ensuring configuration consistency between creation and modification flows.

14. **Inline route editing** — Route cards support in-place editing via a unified form component, eliminating the need for a separate edit dialog or page navigation.

15. **Per-route fallback models** — Each gateway route can define its own ordered fallback model list via the `TagInput` component. When the primary model hits rate limits, the adapter tries these models in order. This replaces the agent-level `fallbackModels` field when gateway is active, providing finer per-route control.

### Architecture

```mermaid
flowchart TD
    HB["Heartbeat Service"] --> BYPASS{"agent.metadata.gatewayBypass?"}
    BYPASS -->|"true"| DEFAULT["Agent default adapter"]
    BYPASS -->|"false"| GW["Gateway resolveRoute()"]
    GW --> |"Route found"| COMPAT{"Cross-adapter?"}
    GW --> |"No routes / DB error"| DEFAULT
    COMPAT -->|"Yes"| CLEAR["Clear fallbackModels"] --> ADAPT["getServerAdapter(gatewayAdapter)"]
    COMPAT -->|"No"| ADAPT
    DEFAULT --> EXEC["adapter.execute()"]
    ADAPT --> EXEC
    EXEC --> |"Success"| REC_OK["recordSuccess(routeId, tokens)"]
    EXEC --> |"Failure/Timeout"| REC_FAIL["recordFailure(routeId)"]
    REC_OK --> COUNTER["Update rate counters + DB upsert"]
    REC_OK --> CB_RESET["Reset circuit breaker"]
    REC_FAIL --> COUNTER
    REC_FAIL --> CB_CHECK{"failureCount ≥ threshold?"}
    CB_CHECK --> |"Yes"| CB_OPEN["Circuit → OPEN (skip route)"]
    CB_CHECK --> |"No"| CB_INC["Increment failure count"]
    CB_OPEN --> PERSIST["persistCircuitState (atomic upsert)"]
    CB_INC --> PERSIST
```

### Review-Round Fixes

All findings from 4 rounds of Greptile AI code review and latest manual testing have been resolved:

| # | Issue | Severity | File | Fix |
|---|---|---|---|---|
| 1 | `loadPersistedCircuitStates` never called — circuit state lost on restart | P1 | `index.ts` | Called during server startup after DB init |
| 2 | `agentId` query param ignored in route listing | P1 | `routes/gateway.ts`, `services/gateway.ts` | Passed through and filtered in `listRoutes` |
| 3 | `persistAdapterManagedRuntimeServices` uses wrong adapter type after routing | P1 | `heartbeat.ts` | Changed to `effectiveAdapterType` |
| 4 | `agentId` not validated as belonging to company — cross-company scoping | P1 Security | `routes/gateway.ts` | Added `assertAgentOwnership()` guard in POST and PATCH |
| 5 | Agent deletion FK violation when gateway routes exist | P0 | `services/agents.ts` | Added `tx.delete(gatewayRoutes)` before agent delete |
| 6 | Company deletion FK violation when gateway routes exist | P0 | `services/companies.ts` | Added `tx.delete(gatewayRoutes)` before agents delete |
| 7 | `gateway_usage_counters` table defined but never written to | P2 | `services/gateway.ts` | Added `persistUsageCounter()` upsert in `recordSuccess`/`recordFailure` |
| 8 | `half_open` circuit breaker test doesn't cover actual code path | P2 | `gateway-service.test.ts` | Rewrote with `vi.spyOn(Date, 'now')` to exercise real `open→half_open→open` transition |
| 9 | `persistCircuitState` race condition — concurrent writes can lose state | P2 | `services/gateway.ts` | Replaced read-then-write with atomic `INSERT ... ON CONFLICT DO UPDATE` upsert |
| 10 | `gateway_usage_counters` rows never purged — unbounded table growth | P2 | `services/gateway.ts`, `index.ts` | Added `purgeStaleUsageCounters()` (7-day TTL), called hourly via tick counter |
| 11 | Wrong Cursor adapter type `cursor_local` → silently routes to processAdapter | P1 | `GatewayRouting.tsx` | Changed to `"cursor"` matching `builtin-adapter-types.ts` |
| 12 | Gateway DB failure crashes all heartbeat runs (no error isolation) | P1 | `heartbeat.ts` | Wrapped `resolveRoute` in `.catch()` with `null` fallback |
| 13 | `open → half_open` transition not persisted — restarts reset cooldown timer | P2 | `services/gateway.ts` | Pass `db` to `isRouteAvailable`, `void persistCircuitState()` on transition |
| 14 | `purgeStaleUsageCounters` runs on every heartbeat tick (~30-60s) | P2 | `index.ts` | Throttled to once per ~hour via tick counter (`GATEWAY_PURGE_EVERY_N_TICKS`) |
| 15 | `recordFailure` unguarded — DB error misclassifying timed-out runs | P1 | `heartbeat.ts` | Wrapped `recordFailure` in `.catch()` and added error log |
| 16 | Missing route-level adapter config (env vars, extra args, etc.) | Feature | `gateway.ts`, `schema`, `UI` | Added `adapterConfigOverrides` JSONB column + UI config section |
| 17 | Gateway route form hardcodes adapter types and duplicates env var editor | P2 | `GatewayRouting.tsx`, `AgentConfigForm.tsx` | Replaced with shared `AdapterTypeDropdown` + `EnvVarEditor` imports; added `agentsApi.adapterModels()` for model hints |
| 18 | Agent-specific routes block company-wide fallbacks when all exhausted | P2 | `services/gateway.ts`, `gateway-service.test.ts` | Merged agent + company routes into single candidate pool; rewrote test to verify fallback behavior |
| 19 | Agent routes not preferred over company-wide at same priority | P2 | `services/gateway.ts`, `gateway-service.test.ts` | Added `agentSpecific` filter within each priority tier; added test for same-priority preference |
| 20 | `paperclip.sh` exposes production IP and placeholder auth secret | P1 Security | `paperclip.sh`, `.gitignore` | File removed from git index (`git rm --cached`); added to `.gitignore` permanently |

### Automated

```bash
pnpm -r typecheck   # ✅ All packages pass
pnpm test:run        # ✅ 954 passed, 1 skipped (1 pre-existing env failure in workspace-runtime)
pnpm build           # ✅ Build succeeds
pnpm db:generate     # ✅ Migration 0050_graceful_speed.sql generated
```

### Gateway Unit Tests (20/20 passed)

```
✓ resolveRoute → returns null when no routes exist
✓ resolveRoute → returns the highest-priority route
✓ resolveRoute → prefers agent-specific routes but falls back to company-wide when exhausted
✓ resolveRoute → prefers agent-specific routes over company-wide at the same priority
✓ resolveRoute → skips routes that have exhausted their request quota
✓ resolveRoute → skips routes that have exhausted their token quota
✓ resolveRoute → returns null when all routes are exhausted
✓ recordSuccess → increments token and request counters
✓ recordSuccess → resets circuit breaker on success after failure
✓ recordFailure → increments failure count and opens circuit at threshold
✓ recordFailure → does nothing if circuit breaker is disabled
✓ recordFailure → re-opens circuit immediately on half_open failure
✓ resetCircuit → resets an open circuit to closed
✓ getRouteHealth → returns clean health for unknown route
✓ getRouteHealth → accumulates usage across multiple successes
✓ circuit breaker integration → skips open circuit, uses lower-priority fallback
✓ edge cases → handles routes with zero weight gracefully
✓ edge cases → handles multiple routes at same priority with weighted selection
✓ edge cases → returns timeoutSec from route decision when set
✓ edge cases → returns null timeoutSec when not configured
```

### Manual Testing

1. **Gateway CRUD**: Created routes via API and UI, verified list/update/delete operations
2. **Priority routing**: Configured 3 routes (P10 Gemini, P5 Codex, P1 OpenCode), verified agents use highest-priority available route
3. **Circuit breaking**: Observed Gemini rate-limit failures causing `recordFailure()` to increment, confirmed circuit opens at threshold
4. **Cross-adapter safety**: Verified that when gateway routes an OpenCode agent to Gemini, the `fallbackModels` (opencode/* models) are cleared — preventing the `ModelNotFoundError (404)` that previously crashed runs
5. **Per-agent bypass**: Toggled gateway On/Off for individual agents via UI, confirmed heartbeat respects the `metadata.gatewayBypass` flag
6. **Light/dark theme**: Verified all gateway UI elements (execution chain banner, cross-adapter note) are legible in both themes
7. **Backward compatibility**: Agents without gateway routes behave identically to before — no config changes needed
8. **Deployment safety**: Verified that gateway resolution gracefully degrades to default adapter when DB query fails (`.catch()` fallback)

### Bug Fixed During Testing

**Cross-Adapter Fallback Crash** — When the gateway routed an `opencode_local` agent to `gemini_local`, the Gemini CLI's internal timeout triggered the agent's legacy `fallbackModels` (`opencode/minimax-m2.5-free`). Gemini CLI doesn't know OpenCode model identifiers → `ModelNotFoundError (404)` → run crashes. Fixed by clearing `fallbackModels` when `effectiveAdapterType !== agent.adapterType`.

## Risks

- **In-memory rate limiting** — Rate limit counters are not shared across server instances. For single-instance dev/self-hosted deployments this is fine. Production multi-instance setups would need Redis-backed counters. Documented in `doc/GATEWAY.md`.
- **Circuit breaker tuning** — Default threshold (3 failures) and cooldown (60s) work for most cases but may need per-vendor tuning. Configurable per-route via API.
- **No intra-run retry** — Gateway fallback is cross-run, not intra-run. If a route fails mid-execution, the current run completes as failed; the next run will skip the tripped circuit. Intra-run retry would require deeper adapter architecture changes.
- **Metadata field usage** — `gatewayBypass` is stored in the agent's `metadata` JSONB column. This is a low-risk approach (no migration), but future features using `metadata` should be aware of this key.
- **Migration safety** — Migrations `0053` and `0054` are additive (CREATE TABLE + ADD COLUMN only), no existing tables are altered. Safe to apply on running systems. Migration numbers were bumped from 0050/0051 to 0053/0054 after resolving merge conflicts with upstream.

## Model Used

- **Provider**: Google DeepMind (Antigravity)
- **Model**: Claude Opus 4.6, extended thinking mode
- **Context**: 1M token context window
- **Capabilities**: Tool use (file editing, command execution, web search), agentic code generation with multi-file coordination
- **Human oversight**: All changes reviewed and tested by the human operator before committing

## Checklist

- [x] I have included a thinking path that traces from project context to this change
- [x] I have specified the model used (with version and capability details)
- [x] I have run tests locally and they pass
- [x] I have added or updated tests where applicable (`server/src/__tests__/gateway-service.test.ts` — 20 tests)
- [x] If this change affects the UI, I have included before/after screenshots
- [x] I have updated relevant documentation to reflect my changes (`doc/GATEWAY.md`)
- [x] I have considered and documented any risks above
- [x] I will address all Greptile and reviewer comments before requesting merge
- [x] **All 20 review findings have been resolved** (see Review-Round Fixes table above)

---

**Files changed**: 32 (13 new, 19 modified) | **Squashed commits** | **~27,000+ / -300 lines**

---

image

image

image

image

image

### CODE REVIEW

File: server/src/services/gateway.ts [hunk 1]
```diff
@@ -0,0 +1,336 @@
...
+        openedAt: row.openedAt,
+      });
+    }
+    if (rows.length > 0) {
+      logger.info({ count: rows.length }, "Loaded persisted gateway circuit breaker states");
+    }
+  } catch (err) {
+    logger.warn({ err }, "Failed to load persisted gateway circuit states");
+  }
+}
```
[the remark this answers is not in the document]
Comment (kinglegendzzh) [reply server/src/services/gateway.ts:381]: Fixed in f2d60bde. `loadPersistedCircuitStates` is now called during server startup in `server/src/index.ts` (right after heartbeat orphan recovery), ensuring open circuits survive restarts:

```typescript
// Restore gateway circuit breaker states from DB so open circuits survive restarts
void loadPersistedCircuitStates(db as any).catch((err) => {
  logger.error({ err }, "startup gateway circuit state recovery failed");
});
```

File: server/src/routes/gateway.ts [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+    const companyId = req.params.companyId as string;
+    assertCompanyAccess(req, companyId);
+
+    const routes = await gw.listRoutes(db, companyId);
+    const enriched = routes.map((route) => ({
+      ...route,
+      health: gw.getRouteHealth(route.id),
+    }));
+    res.json(enriched);
+  });
```
[the remark this answers is not in the document]
Comment (kinglegendzzh) [reply server/src/routes/gateway.ts:32]: Fixed in f2d60bde. The route handler now reads `req.query.agentId` and passes it to `gw.listRoutes()`. The service function filters to return agent-specific routes + company-wide (agentId=null) routes when an agentId is provided, and returns all routes when omitted.

Good catch — fixed in f2d60bde. Changed `agent.adapterType` to `effectiveAdapterType` so runtime services are tagged with the adapter that actually produced them during cross-adapter gateway routing.

**Comment by kinglegendzzh** [comment]
Fixed in 1727c79e. Added `persistUsageCounter()` helper that performs an upsert 
(`INSERT ... ON CONFLICT DO UPDATE`) into `gateway_usage_counters` on every 
`recordSuccess` and `recordFailure` call. Usage data is now persisted to DB 
for observability while in-memory counters remain the fast-path for route decisions.

Fixed in 1727c79e. The test now properly exercises the full half_open → open path:

1. Opens the circuit by reaching the failure threshold
2. Mocks `Date.now` to advance 61s past the 60s reset window
3. Calls `resolveRoute()` which transitions the circuit to `half_open` via `isRouteAvailable()`
4. Calls `recordFailure()` and asserts the circuit re-opens immediately

This covers the previously untested `if (circuit.state === 'half_open')` branch in `recordFailure`.

@greptileai  plz check and review 🙏

File: server/src/routes/gateway.ts [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+      quotaRequestsPerMinute: req.body.quotaRequestsPerMinute ?? null,
+      quotaRequestsPerHour: req.body.quotaRequestsPerHour ?? null,
+      quotaRequestsPerDay: req.body.quotaRequestsPerDay ?? null,
+      circuitBreakerEnabled: req.body.circuitBreakerEnabled ?? false,
+      circuitBreakerFailureThreshold: req.body.circuitBreakerFailureThreshold ?? 3,
+      circuitBreakerResetSec: req.body.circuitBreakerResetSec ?? 300,
+      timeoutSec: req.body.timeoutSec ?? null,
+    });
+
+    res.status(201).json({ ...route, health: gw.getRouteHealth(route.id) });
```
[the remark this answers is not in the document]
Comment (kinglegendzzh) [reply server/src/routes/gateway.ts:71]: Fixed in 4b1bd057. Added `assertAgentOwnership()` helper that validates the agentId 
belongs to the target company via a DB lookup. Called in both POST (create) and 
PATCH (update) handlers before persisting. Returns 400 "agentId does not belong 
to this company" if the check fails.

@greptileai plz check and review

File: ui/src/pages/GatewayRouting.tsx [hunk 4]
```diff
@@ -0,0 +1,583 @@
...
+          <input className={inputCls} value={name} onChange={(e) => setName(e.target.value)} placeholder="e.g. Primary GPT" />
+        </div>
+        <div>
+          <div className={labelCls}>Adapter Type</div>
+          <select className={inputCls} value={adapterType} onChange={(e) => setAdapterType(e.target.value)}>
+            <option value="opencode_local">OpenCode</option>
+            <option value="gemini_local">Gemini CLI</option>
+            <option value="claude_local">Claude Code</option>
+            <option value="codex_local">Codex</option>
+            <option value="cursor_local">Cursor</option>
```
[the remark this answers is not in the document]
Comment (kinglegendzzh) [reply ui/src/pages/GatewayRouting.tsx]: Fixed. Changed `cursor_local` → `cursor` to match the registered adapter type 
in `builtin-adapter-types.ts`. The previous value would have silently fallen 
back to processAdapter.

File: server/src/services/heartbeat.ts [hunk 5]
```diff
@@ -2856,26 +2857,75 @@ export function heartbeatService(db: Db) {
...
 
-      const adapter = getServerAdapter(agent.adapterType);
+      // --- Gateway routing: resolve cross-adapter route if configured ---
+      // Per-agent gateway bypass: if the agent has metadata.gatewayBypass = true,
+      // skip gateway resolution entirely and use the agent's default adapter config.
+      const agentMeta = (agent.metadata ?? {}) as Record<string, unknown>;
+      const gatewayBypassed = agentMeta.gatewayBypass === true;
+      const gatewayDecision = gatewayBypassed
+        ? null
+        : await gatewayService.resolveRoute(db, agent.id, agent.companyId);
```
[the remark this answers is not in the document]
Comment (kinglegendzzh) [reply server/src/services/heartbeat.ts]: 1. resolveRoute DB 错误崩溃所有 heartbeat run | P1 | .catch() + null 降级
2. open → half_open 转换未持久化 | P2 | isRouteAvailable 接收 db，void persistCircuitState()
3. purgeStaleUsageCounters 每个 tick 都执行 | P2 | tick 计数器，仅每 ~1 小时执行一次

File: back/csv_account.csv [hunk 6]
```diff
@@ -0,0 +1,2 @@
+id,account_id,provider_id,user_id,access_token,refresh_token,id_token,access_token_expires_at,refresh_token_expires_at,scope,password,created_at,updated_at
+23K7YgG16IkKWDdXDBPLaVWWo9QeDW7k,MoEIptWrsumBkAdGK9w7UT9gUsAd8N8f,credential,MoEIptWrsumBkAdGK9w7UT9gUsAd8N8f,,,,,,,70e218cbf2862f7c5c8b9a1ec0507a55:8e1b7e4128b90ac3ebed2cf02603ff3577e4441cf11bee584090ecc54dde1bd4a64cdfc1280b31dbd51a91b1609484439b5eb1007098d556a676b8a670028adb,2026-04-04 13:16:55.351+08,2026-04-04 13:16:55.351+08
```
[the remark this answers is not in the document]
Comment (kinglegendzzh) [reply back/csv_account.csv]: its ignored

File: import_data.sh [hunk 7]
```diff
@@ -0,0 +1,70 @@
+#!/bin/bash
+# 清空嵌入式 PG 所有数据，然后从生产环境 CSV 重新导入
+set -e
+
+PSQL="/Applications/Postgres.app/Contents/Versions/latest/bin/psql"
+HOST="127.0.0.1"
+PORT="54329"
+USER="paperclip"
+DB="paperclip"
+export PGPASSWORD="paperclip"
```
[the remark this answers is not in the document]
Comment (kinglegendzzh) [reply import_data.sh]: its ignored

**Comment by kinglegendzzh** [comment]
@greptileai check and review

File: paperclip.sh [hunk 8]
```diff
@@ -0,0 +1,145 @@
...
+# =========================================================
+# Paperclip 终极守护脚本 (专为阿里云+宝塔 www 用户定制)
+# =========================================================
+
+# --- 全局配置 ---
+PROJECT_DIR="/data/apps/paperclip-prod"
+SERVER_DIR="$PROJECT_DIR/server"
+NODE_BIN="/www/server/nodejs/v24.14.1/bin"
+WWW_HOME="/tmp/www_home"
+LOG_FILE="$PROJECT_DIR/paperclip.log"
```
[the remark this answers is not in the document]
Comment (kinglegendzzh) [reply paperclip.sh]: fixed it

**Review by njarun — APPROVED** [review]

**Review by kinglegendzzh — COMMENTED** [review]
👀
